docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,312 @@
+  parameter is ``None`` which when provided, runs coala without a cache.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+* The new bears always have to constructed with ``section`` and ``file_dict``
+  as parameters, this is important as it makes the new dependency system
+  possible.
+* Bear writers need to keep this in mind while writing their ``__init__``
+  methods, default parameters are allowed as many as wanted but the same isn't
+  true for positional parameters (default parameters are discouraged, since the
+  dependency system automatically instantiates bears, which might lead to
+  unwanted side-effects).
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files. Bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs.
+
+Task Objects
+------------
+
+These are the representation of tasks performed by bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the ``analyze`` (even to the ``execute_task``) function.
+The task objects are instrumental to the caching mechanism as their hash values
+are stored in the cache along with the bear results and are looked up during
+each coala run to fetch the results. Lets say we have a FileBear that looks like
+this:
+
+::
+    class SomeFileBear(FileBear):
+
+    def analyze(self, file, filename, filename_prefix: str='',
+                filename_suffix: str=''):
+        yield "Some kind of analysis"
+
+Its corresponding task object would like the one below.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+    ]
+
+Improved dependency system
+--------------------------
+
+The NextGen-Core introduces a better dependency management system than the
+one used by the old core. It features the following improvements:
+
+* A class ``DependencyTracker`` manages this task of dependency management.
+  Dependencies are added and resolved by this class and it checks for
+  circular dependencies.
+* Dependency relations between two objects are tracked using a directed
+  graph. When two objects (or nodes) are connected with a directed edge they
+  form a dependency relation.
+
+Eventhough bears still have to pass ``Result`` instances to communicate with
+coala, it is now possible to pass arbitrary python objects. This has its
+benefits in case of dependency bears since they can pass any arbitrary
+object instead of just ``Result`` objects.
```

Comment:
Maybe rephrase:
Dependency bears benefit from this ...

And maybe a reason is good:
> ... to tailor their passed data to their needs without being bound to `Result` objects.

or similar (that sentence should be improved)